coremltools/converters/nnssa/coreml/graph_pass/mlmodel_passes.py

```diff
@@ -244,27 +245,49 @@ def remove_redundant_transposes(spec):
     that compose to the identity.
     """
 
+    def _get_input_output_hashmap(nn_layers):
+        # Hashmap for layers
+        output_to_layers = {}
+        for layer in nn_layers:
+            for input in layer.input:
+                if not input in output_to_layers:
+                    output_to_layers[input] = [layer]
+                else:
+                    output_to_layers[input].append(layer)
+
+        input_to_layer = {}
+        for layer in nn_layers:
+            for output in layer.output:
+                assert(output not in input_to_layer)
+                input_to_layer[output] = layer
+
+        return input_to_layer, output_to_layers
+
     def _delete_layers(nn_spec, layers_to_delete):
         """
         Given a neural network spec and pairs of transposes to remove, rewire
         the network to bypass those transposes and remove them from the spec.
         """
         nn_layers = nn_spec.layers
+        input_to_layer, output_to_layers = _get_input_output_hashmap(nn_layers)
+
         # First pass: rewire layers to bypass those that will be deleted.
-        for _layer_pair in layers_to_delete:
-            for _nn_layer in nn_layers:
-                if _nn_layer in _layer_pair:
-                    # Skip the layers we're going to delete.
-                    continue
-                if _layer_pair[1].name in _nn_layer.input:
-                    # This layer has one of the deleted as an input. Replace it
-                    # with the deleted layer's input.
-                    idx = [i for i,n in enumerate(_nn_layer.input) if n == _layer_pair[1].name][0]
-                    _nn_layer.input[idx] = _layer_pair[0].input[0]
+        for layers in layers_to_delete:
+            start_layer = layers[0]
+            end_layer = layers[-1]
+
+            # Replace children's input by layer_start's input
+            children = output_to_layers[end_layer.output[0]]
+            for child in children:
+                idx = [i for i, input in enumerate(child.input) if input == end_layer.output[0]]
+                assert(len(idx) == 1)
+                idx = idx[0]
+                child.input[idx] = start_layer.input[0]
+
         # Second pass: delete the layers.
-        for _layer_pair in layers_to_delete:
-            nn_layers.remove(_layer_pair[0])
-            nn_layers.remove(_layer_pair[1])
+        for layers in layers_to_delete:
+            for layer in layers:
+                nn_layers.remove(layer)
 
     def _find_redundant_transposes(nn_spec):
```

Comment:
update the comment of this method which is no longer valid after this change.

Reply:
good!